geoopt/manifolds/stereographic/math.py

```diff
@@ -0,0 +1,2018 @@
+r"""
+:math:`\kappa`-Stereographic math module.
+
+The functions for the mathematics in gyrovector spaces are taken from the
+following resources:
+
+    [1] Ganea, Octavian, Gary Bécigneul, and Thomas Hofmann. "Hyperbolic
+           neural networks." Advances in neural information processing systems.
+           2018.
+    [2] Bachmann, Gregor, Gary Bécigneul, and Octavian-Eugen Ganea. "Constant
+           Curvature Graph Convolutional Networks." arXiv preprint
+           arXiv:1911.05076 (2019).
+    [3] Skopek, Ondrej, Octavian-Eugen Ganea, and Gary Bécigneul.
+           "Mixed-curvature Variational Autoencoders." arXiv preprint
+           arXiv:1911.08411 (2019).
+    [4] Ungar, Abraham A. Analytic hyperbolic geometry: Mathematical
+           foundations and applications. World Scientific, 2005.
+    [5] Albert, Ungar Abraham. Barycentric calculus in Euclidean and
+           hyperbolic geometry: A comparative introduction. World Scientific,
+           2010.
+"""
+
+import functools
+import torch.jit
+from typing import List, Optional
+from ...utils import list_range, drop_dims, sign, clamp_abs
+
+
+@torch.jit.script
+def tanh(x):
+    return x.clamp(-15, 15).tanh()
+
+
+@torch.jit.script
+def artanh(x: torch.Tensor):
+    x = x.clamp(-1 + 1e-15, 1 - 1e-15)
+    return (torch.log(1 + x).sub(torch.log(1 - x))).mul(0.5)
+
+
+@torch.jit.script
+def arsinh(x: torch.Tensor):
+    return (x + torch.sqrt(1 + x.pow(2))).clamp_min(1e-15).log().to(x.dtype)
+
+
+@torch.jit.script
+def abs_zero_grad(x):
+    # this op has derivative equal to 1 at zero
+    return x * sign(x)
+
+
+@torch.jit.script
+def tan_k_zero_taylor(x: torch.Tensor, k: torch.Tensor, order: int = -1):
+    if order == 0:
+        return x
+    k = abs_zero_grad(k)
+    if order == -1 or order == 5:
+        return (
+            x
+            + 1 / 3 * k * x ** 3
+            + 2 / 15 * k ** 2 * x ** 5
+            + 17 / 315 * k ** 3 * x ** 7
+            + 62 / 2835 * k ** 4 * x ** 9
+            + 1382 / 155925 * k ** 5 * x ** 11
+            # + o(k**6)
+        )
+    elif order == 1:
+        return x + 1 / 3 * k * x ** 3
+    elif order == 2:
+        return x + 1 / 3 * k * x ** 3 + 2 / 15 * k ** 2 * x ** 5
+    elif order == 3:
+        return (
+            x
+            + 1 / 3 * k * x ** 3
+            + 2 / 15 * k ** 2 * x ** 5
+            + 17 / 315 * k ** 3 * x ** 7
+        )
+    elif order == 4:
+        return (
+            x
+            + 1 / 3 * k * x ** 3
+            + 2 / 15 * k ** 2 * x ** 5
+            + 17 / 315 * k ** 3 * x ** 7
+            + 62 / 2835 * k ** 4 * x ** 9
+        )
+    else:
+        raise RuntimeError("order not in [-1, 5]")
+
```

Comment:
Can loop be use here to simplify the code?

Reply:
I thought about it, don't now how much readability it adds

Reply:
Well maybe dynamic order with numerator/denominator ([A036279](https://oeis.org/A036279)/[A002430](https://oeis.org/A002430) ) sequence in list somewhere in globals?

Reply:
I consider rewriting this to Cpp at some point. I'd prefer to leave this for now without overcomplication.

Reply:
And I don't think it is important to have a very long tailor series